### ISSUE TITLE
chainer codes that correctly run in the windows but in ubuntu has error

### ISSUE THREAD

**Issue by erfangian** [chainer/chainer#3288]
i have the chainer code that on the windows correctly run .
now i want to run this code on the ubuntu server but i get this error ..

chainer.utils.type_check.InvalidType:
Invalid operation is performed in: EmbedIDFunction (Forward)

Expect: in_types[0].dtype == <class 'numpy.int32'>
Actual: int64 != <class 'numpy.int32'>

**Comment by niboshi** [comment]
Thank you for reporting!
Could you provide more information such as:
* Code
* Stack trace
* Chainer version

**Comment by erfangian** [comment]
the code is Available in  https://github.com/soskek/convolutional_seq2seq
in this code first using link.EmbedID and create a EmbedID . then when using this EmbedID getting this error 
that related to dtype . this code correctly run on the windows  ... 
 chainer version: 2.0.2

**Comment by erfangian** [comment]
please help me ... how can i fix this error

**Comment by niboshi** [comment]
What happens if you rewrite `numpy.array(ids, 'i')` as `numpy.array(ids, numpy.int32)`?
(at https://github.com/soskek/convolutional_seq2seq/blob/master/europal.py#L74)

**Comment by erfangian** [comment]
thanks ... i change  numpy.array(ids , ..) but no change in the run and i get same error

**Comment by soskek** [comment]
Can you show the full trace of the error including lines before what you show?
```
(I want to see all lines before this)

chainer.utils.type_check.InvalidType:
Invalid operation is performed in: EmbedIDFunction (Forward)

Expect: in_types[0].dtype == <class 'numpy.int32'>
Actual: int64 != <class 'numpy.int32'>
```

**Comment by erfangian** [comment]
yes Of course:

Original training data size: 66719
Filtered training data size: 59500
len(sourse): 9119
len(target): 9119
Number of iter/epoch = 1859
start training
Traceback (most recent call last):
  File "seq2seq.py", line 233, in <module>
    main()
  File "seq2seq.py", line 229, in main
    trainer.run()
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/training/trainer.py", line 296, in run
    update()
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/training/updater.py", line 223, in update
    self.update_core()
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/training/updater.py", line 234, in update_core
    optimizer.update(loss_func, *in_arrays)
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/optimizer.py", line 527, in update
    loss = lossfun(*args, **kwds)
  File "/home/erahmani/convolution/net.py", line 206, in __call__
    ey_block = sentence_block_embed(self.embed_y, y_in_block)
  File "/home/erahmani/convolution/net.py", line 20, in sentence_block_embed
    e = embed(x.reshape((batch*length, )))
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/links/connection/embed_id.py", line 53, in __call__
    return embed_id.embed_id(x, self.W, ignore_label=self.ignore_label)
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/functions/connection/embed_id.py", line 111, in embed_id
    return EmbedIDFunction(ignore_label=ignore_label)(x, W)
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/function.py", line 187, in __call__
    self._check_data_type_forward(in_data)
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/function.py", line 290, in _check_data_type_forward
    self.check_type_forward(in_type)
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/functions/connection/embed_id.py", line 20, in check_type_forward
    x_type.ndim >= 1,
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/utils/type_check.py", line 514, in expect
    expr.expect()
  File "/home/erahmani/.local/lib/python3.5/site-packages/chainer/utils/type_check.py", line 472, in expect
    '{0} {1} {2}'.format(left, self.inv, right))
chainer.utils.type_check.InvalidType:
Invalid operation is performed in: EmbedIDFunction (Forward)

Expect: in_types[0].dtype == <class 'numpy.int32'>
Actual: int64 != <class 'numpy.int32'>

**Comment by erfangian** [comment]
i understand that a few element of the x in the method "sentence_block_embed" in "net.py" has data type int64 .
i try to convert this data type into int32 in this line ( e = embed(x.reshape((batch * length, ))))
but the 32*42 = 1344 in int64 when convert to int32 has 2688 element and get this error that the dimension is not correct

**Comment by erfangian** [comment]
i fix this error by using a x.astype(np.int32) in lines that i gets error 
in this code ....
in "sentences_block_embed" in "net.py" for x
and in the class "seq2seq" for "concat_y_out_block"

**Comment by erfangian** [comment]
Thank you for your reply

**Comment by soskek** [comment]
So, you fix it to `e = embed(x.astype(np.int32).reshape((batch * length, )))` ? Then, I'll follow it.
I have not yet understood what happens. Anyway, congratulations!

**Comment by erfangian** [comment]
yes ..
x_re=x.reshape((batch*length,,))
print(x_re.dtype)   ## int64
x_re=x_re.astype(np.int32)
print(x_re.dtype)  ## int32
e-embed(x_re)
if you see the error this told the EmbedID expect int32 but the x has data type int64 ..
first i using the x.dtype="int32" but this changing the dimension ... x.astype(np.int32) keep the dimension without any change ... 
now i run this code bet it is very very slow on the NVIDIA Gt 720 ..

**Comment by niboshi** [comment]
Just to be sure, could you give the output of this code?
```
import numpy
print(numpy.array(3, 'i').dtype)
```

**Comment by beam2d** [comment]
I came up with the situation that this kind of thing happens (though I'm not sure if it's actually happening in the current issue). When a list of Python int values are converted to ndarray without specifying dtype, it will have `int32` on Windows while have `int64` on Linux, which may cause a bug only occurring on Linux. I think it should be fixed by Chainer itself, but anyway you can make a workaround by specifying `dtype='int32'` whenever converting a list of integers to ndarray.

**Comment by kmaehashi** [comment]
This issue has been resolved in #3466.
